src/skadi/sample.cc

```diff
@@ -474,74 +460,136 @@ template <class coords_t> std::vector<double> sample::get_all(const coords_t& co
-  tile_data tile(cache_, TILE_COUNT, false, nullptr);
+  tile_data tile(cache_.get(), TILE_COUNT, false, nullptr);
   for (const auto& coord : coords) {
     auto lon = std::floor(coord.first);
     auto lat = std::floor(coord.second);
     auto index = static_cast<uint16_t>(lat + 90) * 360 + static_cast<uint16_t>(lon + 180);
 
     if (index != curIndex) {
-      tile = cache_->source(index);
       curIndex = index;
+      tile = cache_->source(index);
     }
 
-    double value = NO_DATA_VALUE;
+    double value{get_no_data_value()};
     if (tile) {
       double u = (coord.first - lon) * (HGT_DIM - 1);
       double v = (1.0 - (coord.second - lat)) * (HGT_DIM - 1);
       value = tile.get(u, v);
     }
 
+    if (value == get_no_data_value())
+      value = get_from_remote(coord);
+
     values.emplace_back(value);
   }
+
   return values;
 }
 
-double sample::get_no_data_value() {
-  return NO_DATA_VALUE;
-}
+template <class coord_t> double sample::get_from_cache(const coord_t& coord) {
+  auto tile = cache_->source(get_tile_index(coord));
+  if (!tile)
+    return get_no_data_value();
 
-template <class coord_t> uint16_t sample::get_tile_index(const coord_t& coord) {
-  auto lon = std::floor(coord.first);
-  auto lat = std::floor(coord.second);
-  return static_cast<uint16_t>(lat + 90) * 360 + static_cast<uint16_t>(lon + 180);
+  double u = (coord.first - std::floor(coord.first)) * (HGT_DIM - 1);
+  double v = (1.0 - (coord.second - std::floor(coord.second))) * (HGT_DIM - 1);
+  return tile.get(u, v);
 }
 
-void sample::add_single_tile(const std::string& path) {
-  cache_->cache.front().init(path, format_t::RAW);
+bool sample::store(const std::string& elev, const std::vector<char>& raw_data) {
+  // data_source never changes so we do not lock it. it is set only in sample constructor
+  auto fpath = cache_->data_source + elev;
+  if (filesystem::exists(fpath))
+    return true;
+
+  auto data = cache_item_t::parse_hgt_name(elev);
+  if (!data || data->second == format_t::UNKNOWN)
+    return false;
+
+  // cache size is always the same.
+  if (data->first >= cache_->size())
+    return false;
+
+  // thread-safe by implementation
+  if (!filesystem::save(fpath, raw_data))
+    return false;
+
+  std::lock_guard<std::mutex> _(cache_lck);
+  return cache_->insert(data->first, fpath, data->second);
 }
 
-std::string sample::get_hgt_file_name(uint16_t index) {
-  auto x = (index % 360) - 180;
-  auto y = (index / 360) - 90;
+template <class coord_t> double sample::get_from_remote(const coord_t& coord) {
+  if (url_.empty() || !remote_loader_)
+    return get_no_data_value();
 
-  std::string name(y < 0 ? "/S" : "/N");
-  y = std::abs(y);
-  if (y < 10) {
-    name.push_back('0');
+  auto index = get_tile_index(coord);
+  auto elev = get_hgt_file_name(index);
+  auto uri = baldr::make_single_point_url(url_, elev, remote_path_);
+
+  {
+    std::shared_lock<std::shared_timed_mutex> _(st_lck);
+    if (st_.count(uri))
```

Comment:
It's probably redundant. `If tile exists on filesystem`-check is enough. I mean if we invoke `get_from_remote` function we're sure that we need to fetch this tile

Reply:
the idea was to reduce the number of times we are trying to access filesystem but as discussed yesterday i will remove this as the whole method is going to change.